Haskell/Haskell.sublime-syntax

```diff
@@ -5,68 +5,142 @@ name: Haskell
 file_extensions:
   - hs
 scope: source.haskell
+
+variables:
+  # In case this regex seems overly general,
+  # note that Haskell permits the definition of new operators
+  # which can be nearly any string of punctuation characters,
+  # such as $%^&*.
+  operator_char: '[*|!%&$?~+:\-.=</>\\]'
+  operator_infix: '{{operator_char}}+'
+  operator_parens: '(?:{{operator_infix}}|,+)'
+  module_name: (?:[A-Z][A-Za-z._']*)
+  escape_chars: |-
+    (?x:NUL|SOH|STX|ETX|EOT|ENQ|ACK|BEL|BS|HT|LF|VT|FF|CR|SO|SI
+      |DLE|DC1|DC2|DC3|DC4|NAK|SYN|ETB|CAN|EM|SUB|ESC|FS|GS|RS|US|SP|DEL
+      |[abfnrtv\\\"'\&])
+  escape_sequence: |-
+    (?x:
+      (\\{{escape_chars}})   # Escapes
+      | (\\[0-9]+)           # Decimal Escapes
+      | (\\o[0-7]+)          # Octal Escapes
+      | (\\x[0-9A-Fa-f]+)    # Hexadecimal Escapes
+      | (\^[A-Z@\[\]\\\^_])  # Control Chars
+    )
+
 contexts:
+  prototype:
+    - include: comment
+    - include: preprocessor
+  
   main:
-    - match: "(`)[a-zA-Z_']*?(`)"
-      comment: "In case this regex seems unusual for an infix operator, note that Haskell allows any ordinary function application (elem 4 [1..10]) to be rewritten as an infix expression (4 `elem` [1..10])."
-      scope: keyword.operator.function.infix.haskell
-      captures:
-        1: punctuation.definition.entity.haskell
-        2: punctuation.definition.entity.haskell
+    - include: statement
+    - include: expression
+
+  statement:
+    - include: declaration
+    - include: import
+
+  expression:
+    - include: keyword
+    - include: operator
+    - include: constant
+    - include: string
+    - include: splice
+    - include: group
+    - include: list
+    - include: ident
+    - match: ','
+      scope: punctuation.separator.comma.haskell
+
+  group:
+    - match: \(
+      scope: punctuation.section.group.begin.haskell
+      push:
+        - meta_scope: meta.group.haskell
+        - match: \)
+          scope: punctuation.section.group.end.haskell
+          pop: true
+        - include: expression
+
+  list:
+    - match: \[
+      scope: punctuation.section.sequence.begin.haskell
+      push:
+        - meta_scope: meta.sequence.haskell
```

Comment:
Same as above, we shouldn't need that scope, it matches functions that shouldn't be matched with that scope. Those are supposed to be built-in prelude function names (see on the left, where I'm highlighting them manually)

![image](https://user-images.githubusercontent.com/287532/74486629-b4f59400-4ebd-11ea-8f8c-116f380106c4.png)

Reply:
There is no occurance of `[` in your screenshot here.

Reply:
Yes, that screenshot goes up on that other comment about groups. But it's more or less the same issue. [If you'd like to match names coming from the prelude package](https://github.com/FichteForks/Packages/pull/4/files), for example, you'll see that matching groups and sequences kind of gets in the way...